Задание 2.4

В аптеку поступают партии лекарств. Их названия находятся в списке names, количество единиц товара находится в списке counts.

Например:

names = ['chlorhexidine', 'cyntomycin', 'afobazol'] counts = [15, 18, 7]

Напишите функцию create_medications(names, counts), которая создает Series medications, индексами которой являются названия лекарств names, а значениями - их количество в поставке counts.

А также напишите функцию get_percent(medications, name), которая возвращает долю количества товара с именем name от общего количества товаров в поставке в процентах.

Примечание. Не забудьте ипортировать библиотеки.

Пример:


names=['chlorhexidine', 'cyntomycin', 'afobazol']
counts=[15, 18, 7]
medications = create_medications(names, counts)

print(get_percent(medications, "chlorhexidine")) #37.5

In [55]:
import pandas as pd

def create_medications(names, counts):
    """
    Создает Series, где названия лекарств - это индексы, а их количество - значения.
    """
    return pd.Series(counts, index=names)

def get_percent(medications, name):
    """
    Возвращает процентное соотношение количества указанного лекарства
    к общему количеству всех лекарств в поставке.
    """
    if name not in medications:
        return 0  # Если лекарства нет в списке, возвращаем 0%
    total = medications.sum()
    return (medications[name] / total) * 100

# Пример использования:
names = ['chlorhexidine', 'cyntomycin', 'afobazol']
counts = [15, 18, 7]
medications = create_medications(names, counts)

print(get_percent(medications, "chlorhexidine"))  # 37.5


37.5


Задание 3.5 

Вы работаете аналитиком в компании ScienceYou. Ваша задача — проанализировать чистую прибыль.

Доходы (income), расходы (expenses) и годы (years), соответствующие им, предоставлены вам в виде списков.

Например:


income = [478, 512, 196]
expenses = [156, 130, 270]
years = [2018, 2019, 2020]
Создайте функцию create_companyDF(income, expenses, years), которая возвращает DataFrame, составленный из входных данных со столбцами Income и Expenses и индексами, соответствующими годам рассматриваемого периода.

Пример такого DataFrame представлен ниже.


    Income  Expenses
2018    478     156
2019    512     130
2020    196     270
Также напишите функцию get_profit(df, year), которая возвращает разницу между доходом и расходом, записанными в таблице df, за год year.

Примечание. Если информация за запрашиваемый год не указана в вашей таблице, вам необходимо вернуть None.

In [56]:
import pandas as pd

def create_companyDF(income, expenses, years):
    data = {'Income': income, 'Expenses': expenses}
    df = pd.DataFrame(data, index=years)
    return df

def get_profit(df, year):
    if year in df.index:
        return df.loc[year, 'Income'] - df.loc[year, 'Expenses']
    return None

# Пример использования
income = [478, 512, 196]
expenses = [156, 130, 270]
years = [2018, 2019, 2020]

df = create_companyDF(income, expenses, years)
print(df)

profit_2019 = get_profit(df, 2019)
print(f"Разница между доходом и расходом за 2019 год: {profit_2019}")


      Income  Expenses
2018     478       156
2019     512       130
2020     196       270
Разница между доходом и расходом за 2019 год: 382


Данные, с которыми мы будем работать — это обработанный датасет об объектах недвижимости в Мельбурне (Австралия) и его пригородах.

Данные представляют собой таблицу, в которой содержится 23 столбца:

index — номер строки
Suburb — наименование пригорода
Address — адрес
Rooms — количество комнат в помещении
Type — тип здания (h — дом, коттедж, вилла, терраса; u — блочный, дуплексный дом; t — таунхаус)
Price — цена помещения
Method — метод продажи 
SellerG — риэлторская компания
Date — дата продажи (в формате день/месяц/год)
Distance — расстояния до объекта от центра Мельбурна 
Postcode — почтовый индекс
Bedroom — количество спален
Bathroom — количество ванных комнат
Car — количество парковочных мест
Landsize — площадь прилегающей территории
BuildingArea — площадь здания
YearBuilt — год постройки
CouncilArea — региональное управление
Lattitude — географическая широта
Longitude — географическая долгота
Regionname — наименование района Мельбурна
Propertycount — количество объектов недвижимости в районе, выставленных на продажу
Coordinates — широта и долгота, объединённые в кортеж

In [57]:
#Прочитаем наши данные о недвижимости из csv-файла и запишем результирующий DataFrame в переменную melb_data:

melb_data = pd.read_csv('data/melb_data.csv', sep=',')
display(melb_data)

index         Suburb           Address  Rooms Type      Price Method  \
0          0     Abbotsford      85 Turner St      2    h  1480000.0      S   
1          1     Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2          2     Abbotsford      5 Charles St      3    h  1465000.0     SP   
3          3     Abbotsford  40 Federation La      3    h   850000.0     PI   
4          4     Abbotsford       55a Park St      4    h  1600000.0     VB   
...      ...            ...               ...    ...  ...        ...    ...   
13575  13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
0        Biggin   3/12/2016       2.5  ...  1.0     202.0         126.0   
1        Biggin   4/02/2016       2.5  ...  0.0     156.0          79.0   
2        Biggin   4/03/2017       2.5  ...  0.0     134.0         150.0   
3        Biggin   4/03/2017       2.5  ...  1.0      94.0         126.0   
4        Nelson   4/06/2016       2.5  ...  2.0     120.0         142.0   
...         ...         ...       ...  ...  ...       ...           ...   
13575     Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576  Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577     Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578   Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579   Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
0         1970.0        Yarra  -37.79960   144.99840   
1         1900.0        Yarra  -37.80790   144.99340   
2         1900.0        Yarra  -37.80930   144.99440   
3         1970.0        Yarra  -37.79690   144.99690   
4         2014.0        Yarra  -37.80720   144.99410   
...          ...          ...        ...         ...   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
0           Northern Metropolitan         4019.0    -37.7996, 144.9984  
1           Northern Metropolitan         4019.0    -37.8079, 144.9934  
2           Northern Metropolitan         4019.0    -37.8093, 144.9944  
3           Northern Metropolitan         4019.0    -37.7969, 144.9969  
4           Northern Metropolitan         4019.0    -37.8072, 144.9941  
...                           ...            ...                   ...  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[13580 rows x 23 columns]

Задание 5.1

In [58]:
#Какова цена объекта недвижимости под индексом 15?
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
price_index_15 = melb_data.loc[15, 'Price']
display(price_index_15)

1310000.0

In [59]:
#Когда был продан объект под индексом 90?
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
price_index_90 = melb_data.loc[90, 'Date']
display(price_index_90)

'10/09/2016'

Задание 5.3

Во сколько раз площадь прилегающей территории, на которой находится здание
с индексом 3521, больше площади участка, на котором находится здание 
с индексом 1690? Ответ округлите до целого числа.

In [ ]:
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
# Извлекаем площадь прилегающей территории (Landsize) для объектов 
# с индексами 3521 и 1690.
landsize_3521 = melb_data.loc[3521, 'Landsize']
landsize_1690 = melb_data.loc[1690, 'Landsize']
# Делим одно значение на другое.
# Округляем результат до целого числа с помощью round().
ratio = round(landsize_3521 / landsize_1690)
display(ratio)

3

Задание 6.1

Дана таблица data = pd.DataFrame([[0,1], [1, 0], [1, 1]], columns=['А', 'B']). Что выведет код print(data.shape)?

Чтобы узнать размер таблицы — количество строк и количество столбцов. Это можно сделать с помощью атрибута shape, который возвращает кортеж с количеством строк и столбцов:

In [64]:
data = pd.DataFrame([[0,1], [1, 0], [1, 1]], columns=['А', 'B'])
print(data.shape)


(3, 2)


Задание 6.2

Какой из нижеперечисленных вариантов кода позволяет получить последние 15 строк DataFrame melb_data?

melb_data.tail(15)

при этом head() верхние (по умолчанию 5)
tail() нижние пять 

Задание 6.3
melb_data.info()
Какой тип данных имеет столбец Coordinates?

 22  Coordinates    13580 non-null  object 
посмотрели с помощью метода info()
Данный метод выводит:

информацию об индексах;
информацию об общем количестве столбцов;
таблицу, в которой содержится информация об именах столбцов (Column), количестве непустых значений (Non-Null Count) в каждом столбце и типе данных столбца (Dtype), количестве столбцов, в которых используется определённый тип данных;
количество оперативной памяти в мегабайтах, которое тратится на хранение данных.

Задание 6.4

Какое утверждение о типе данных object верно?

Это тип данных, которым в Pandas кодируются текстовые или другие нечисловые значения (кортежи, списки и т. д.)

Задание 6.5
melb_data.info()
Сколько пропущенных значений в столбце CouncilArea?
ответ 1369

 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattitude      13580 non-null  float64

 13580-12211= 1369

Задание 6.6
melb_data.info()
Сколько столбцов после преобразования типов имеет целочисленный тип данных?



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  int64  
 12  Bathroom       13580 non-null  int64  
 13  Car            13580 non-null  int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattitude      13580 non-null  float64
 19  Longtitude     13580 non-null  float64
...
 21  Propertycount  13580 non-null  int64  
 22  Coordinates    13580 non-null  object 
dtypes: float64(6), int64(8), object(9)
memory usage: 2.4+ MB

Задание 6.7

Выберите верные утверждения в отношении таблицы melb_data:


Информацию о статистических параметрах столбцов с данными типа object можно получить с помощью метода describe() с параметром include=['object'].


A Наименьшее число объектов расположено в районе Southern Metropolitan
B Мельбурн и его пригороды разделены как минимум на восемь районов
C У некоторых различных адресов в таблице совпадают географические координаты
D В пригороде Reservoir расположено наибольшее число проданных объектов - 359


ни одно утверждение не верно!!!

Задание 6.9

Сколько процентов от общего количества домов составляют таунхаусы (тип объекта — t)?
Ответ округлите до целого числа. Не указывайте знак %.

In [65]:
# Посчитаем общее количество записей в melb_data
total_count = len(melb_data)  # Общее количество объектов
# Вычисляем количество таунхаусов (Type == 't').
townhouse_count = (melb_data['Type'] == 't').sum()  # Количество таунхаусов
# Найти процентное соотношение и округлить до целого числа.
percentage = round((townhouse_count / total_count) * 100)
print(percentage)

8


Задание 7.2

Чему равно максимальное количество домов на продажу в районе (Propertycount)?

In [ ]:
print(melb_data['Propertycount'].max())
#21650.0

Задание 7.3

Чему равно стандартное отклонение (разброс) расстояния от центра города до объекта недвижимости?

In [ ]:
# Distance — расстояния до объекта от центра Мельбурна 

income = melb_data['Distance'].std()
print(round(income))
#6

Задание 7.4

Чему равно отклонение (в процентах) медианного значения площади здания от его среднего значения?

In [ ]:
# BuildingArea — площадь здания

landsize_median = melb_data['BuildingArea'].median() 
landsize_mean =  melb_data['BuildingArea'].mean()

print(round(abs((landsize_median - landsize_mean)/landsize_mean) * 100))
#10

Задание 7.5

Задан ряд чисел [1, 2, 4, 2, 3, 2, 1, 5, 6]. Чему равна мода в данном ряду?

mode - это какое число чаще всего представлено в ряде чисел
у нас это 2 (3 раза)

Задание 7.6

Сколько спален чаще всего встречается в домах в Мельбурне?

In [ ]:
#Вычислим, какое сколько спален чаще всего встречаются
# Bedroom — количество спален
print(melb_data['Bedroom'].mode())
# 0    3.0
# dtype: float64

Данные представляют собой таблицу, в которой содержится 23 столбца:

index — номер строки
Suburb — наименование пригорода
Address — адрес
Rooms — количество комнат в помещении
Type — тип здания (h — дом, коттедж, вилла, терраса; u — блочный, дуплексный дом; t — таунхаус)
Price — цена помещения
Method — метод продажи 
SellerG — риэлторская компания
Date — дата продажи (в формате день/месяц/год)
Distance — расстояния до объекта от центра Мельбурна 
Postcode — почтовый индекс
Bedroom — количество спален
Bathroom — количество ванных комнат
Car — количество парковочных мест
Landsize — площадь прилегающей территории
BuildingArea — площадь здания
YearBuilt — год постройки
CouncilArea — региональное управление
Lattitude — географическая широта
Longitude — географическая долгота
Regionname — наименование района Мельбурна
Propertycount — количество объектов недвижимости в районе, выставленных на продажу
Coordinates — широта и долгота, объединённые в кортеж

Задание 8.1

У скольких объектов недвижимости из таблицы melb_data отсутствуют ванные комнаты?

In [ ]:
#Bathroom — количество ванных комнат
melb_data[melb_data['Bathroom'] == 0].shape[0]
#34

Задание 8.2

Сколько в таблице melb_data объектов недвижимости, которые были проданы риелторской компанией Nelson и стоимость которых составила больше 3 миллионов?

In [ ]:
# SellerG — риэлторская компания

melb_data[(melb_data['SellerG'] == "Nelson") & (melb_data['Price'] > 3000000)].shape[0]
 # 5

Задание 8.3

Какова минимальная стоимость участка без здания (площадь здания равна 0) в таблице melb_data?
Запишите ответ в виде целого числа.

In [ ]:
#BuildingArea — площадь здания
# .min()	Минимальное значение
min_price = melb_data[melb_data['BuildingArea'] == 0]['Price'].min()
print(int(min_price))
#412500

Задание 8.4

Какова средняя цена объектов недвижимости в таблице melb_data с ценой менее одного миллиона, в которых либо количество комнат больше пяти, либо здание моложе 2015 года?
Округлите ответ до целого числа.

In [ ]:
#А теперь более сложный трюк: найдём медианную площадь здания у объектов,
# чья цена выше средней. Для того чтобы оградить наш код от нагромождений,
 
# создадим переменную со средней ценой
# Rooms — количество комнат в помещении
# YearBuilt — год постройки

filtered_data = melb_data[(melb_data['Price'] < 1000000) & 
                          ((melb_data['Rooms'] > 5) | (melb_data['YearBuilt'] > 2015))]

mean_price = round(filtered_data['Price'].mean())
print(mean_price)
#769239

Задание 8.5

В каком районе Мельбурна чаще всего продаются виллы и коттеджи (тип здания — h) с ценой меньше трёх миллионов?

In [104]:
filtered_data = melb_data[(melb_data['Type'] == 'h') & (melb_data['Price'] < 3000000)]

most_common_region = filtered_data['Regionname'].mode()[0]
print(most_common_region)

Northern Metropolitan


gender — пол;
race/ethnicity — раса/этническая принадлежность;
parental level of education — уровень образования родителей;
lunch — какие обеды получал студент во время обучения (standard — платный, free/reduced — бесплатный);
test preparation course — посещал ли студент курсы подготовки к экзаменам (none — не посещал, completed — посещал);
math score, reading score, writing score — баллы по математике, чтению и письму по сто балльной шкале.

In [ ]:
Задание 9.1

Данные о скольких студентах содержатся в таблице?
Ответ 1000

In [16]:
#Прочитаем наши данные о недвижимости из csv-файла и запишем результирующий DataFrame в переменную student_data:

student_data = pd.read_csv('data/students_performance.csv', sep=',')
display(student_data)
#1000 rows × 8 columns

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
0                      none          72             72             74  
1                 completed          69             90             88  
2                      none          90             95             93  
3                      none          47             57             44  
4                      none          76             78             75  
..                      ...         ...            ...            ...  
995               completed          88             99             95  
996                    none          62             55             55  
997               completed          59             71             65  
998               completed          68             78             77  
999                    none          77             86             86  

[1000 rows x 8 columns]

Задание 9.2

Каков балл по письму у студента под индексом 155?

gender — пол;
race/ethnicity — раса/этническая принадлежность;
parental level of education — уровень образования родителей;
lunch — какие обеды получал студент во время обучения (standard — платный, free/reduced — бесплатный);
test preparation course — посещал ли студент курсы подготовки к экзаменам (none — не посещал, completed — посещал);
math score, reading score, writing score — баллы по математике, чтению и письму по сто балльной шкале.

In [17]:
# writing score - баллы по письму по сто балльной шкале.

student_data_155 = student_data.loc[155, 'writing score']
display(student_data_155)
#88

88

Задание 9.3

Сколько суммарно пропущенных значений в таблице?

Их нет!!! Нет пропусков

In [18]:
student_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


Задание 9.4

Сколько столбцов в таблице имеет тип данных object?

Ответ: dtypes: int64(3), object(5)

Задание 9.5

Какой объём памяти (в килобайтах) занимает таблица?
Округлите ответ до целого числа.

Ответ: memory usage: 62.6+ KB (окрглили до 63)

Задание 9.6

Каков у студентов средний балл по математике?
Округлите ответ до целого числа.

Метод	Статистический параметр
.count()	Количество непустых значений
.mean()	Среднее значение
.min()	Минимальное значение
.max()	Максимальное значение
.var()	Дисперсия
.std()	Стандартное отклонение
.sum()	Сумма
.quantile(x)	Квантиль уровня x
.nunique()	Число уникальных значений

In [19]:
# math score, reading score, writing score — баллы по математике, чтению и письму по сто балльной шкале.
# .mean()	Среднее значение
mean_math_score = student_data['math score'].mean()
print(round(mean_math_score))

66


Задание 9.7

Какая расовая группа является самой крупной в учебном заведении?
В качестве ответа введите идентификатор группы (A, B, C, D, E).

In [20]:
#Вычислим, какое число комнат чаще всего представлено на рынке недвижимости:
# Примечание. Метод mode() может быть использован не только с числовыми столбцами,
# но и со столбцами типа object. Так, например, с помощью следующего кода можно 
# найти наиболее распространённое название
print(student_data['race/ethnicity'].mode())

# group C

0    group C
Name: race/ethnicity, dtype: object


Задание 9.8

Каков средний балл по чтению у студентов, которые посещали курсы подготовки к экзаменам?
Округлите ответ до целого числа.

gender — пол;
race/ethnicity — раса/этническая принадлежность;
parental level of education — уровень образования родителей;
lunch — какие обеды получал студент во время обучения (standard — платный, free/reduced — бесплатный);
test preparation course — посещал ли студент курсы подготовки к экзаменам (none — не посещал, completed — посещал);
math score, reading score, writing score — баллы по математике, чтению и письму по сто балльной шкале.

In [21]:
avg_reading_score = round(student_data[student_data['test preparation course'] == 'completed']['reading score'].mean())
print(avg_reading_score)

# 74

74


Задание 9.9

Сколько студентов получили 0 баллов по математике?

In [22]:
# math score, reading score, writing score — баллы по математике, чтению и письму по сто балльной шкале.
student_data[student_data['math score'] == 0].shape[0]

#1

1

Задание 9.10

Проверьте гипотезу: у студентов с оплачиваемым питанием средний балл по математике выше, чем у студентов с льготным питанием.
В качестве ответа напишите наибольший средний балл по математике среди этих групп студентов.
Округлите ответ до целого числа.

In [23]:

mean_standard = round(student_data[student_data['lunch'] == 'standard']['math score'].mean())
mean_free = round(student_data[student_data['lunch'] == 'free/reduced']['math score'].mean())

print(mean_standard) # 70
print(mean_free) # 59

70
59


Задание 9.11

Каков процент студентов, родители которых имеют высшее образование уровня бакалавриата (bachelor's degree)?
Округлите ответ до целого числа.
Не указывайте знак %.

In [24]:
total_students = len(student_data)  # Общее количество студентов
bachelor_students = (student_data['parental level of education'] == "bachelor's degree").sum()  # Количество студентов с родителями-бакалаврами

percentage = round((bachelor_students / total_students) * 100)
print(total_students)
print(percentage)

1000
12


Задание 9.12

Насколько медианный балл по письму у студентов в расовой группе А отличается от среднего балла по письму у студентов в расовой группе C?
Округлите ответ до целого и запишите модуль этого числа.

In [25]:
# writing score — баллы по письму по сто балльной шкале.
# .mean()	Среднее значение
# .median() медиана
median_writing_A = student_data[student_data['race/ethnicity'] == 'group A']['writing score'].median()
mean_writing_C = student_data[student_data['race/ethnicity'] == 'group C']['writing score'].mean()

difference = abs(round(median_writing_A - mean_writing_C))
print(difference)

6


Метод	Статистический параметр
.count()	Количество непустых значений
.mean()	Среднее значение
.min()	Минимальное значение
.max()	Максимальное значение
.var()	Дисперсия
.std()	Стандартное отклонение
.sum()	Сумма
.quantile(x)	Квантиль уровня x
.nunique()	Число уникальных значений

Задание 2.1

Среди приведённых ниже вариантов кода выберите тот, который найдёт квадрат цены объекта недвижимости за наименьшее время и не выдаст ошибку.
Результат должен быть занесён в переменную price_square и представлять собой объект Series:

In [ ]:
price_square = melb_df['Price'] ** 2

Задание 2.2

Задан DataFrame customer_df, содержащий столбцы:
cust_id — идентификатор клиента;
cust_age — возраст клиента (точкой отсчёта возраста считается 2021 год);
cust_sale — персональная скидка клиента;
cust_year_birth — год рождения клиента;
cust_order — сумма заказа клиента.

Какие столбцы не несут полезной информации/дублируют информацию из других столбцов и поэтому могут быть удалены?

В качестве ответа запишите названия этих столбцов по порядку их следования в коде, через запятую, без пробелов и кавычек.

Ответ: number,cust_age

In [27]:
customer_df = pd.DataFrame({
        'number': [0, 1, 2, 3, 4],
        'cust_id': [128, 1201, 9832, 4392, 7472],
        'cust_age': [13, 21, 19, 21, 60],
        'cust_sale': [0, 0, 0.2, 0.15, 0.3],
        'cust_year_birth': [2008, 2000, 2002, 2000, 1961],
        'cust_order': [1400, 14142, 900, 1240, 8430]
    })

# Ответ: number,cust_age

Задание 2.3

Напишите функцию delete_columns(df, col=[]), которая удаляет столбцы из DataFrame и возвращает новую таблицу. Если одного из указанных столбцов не существует в таблице, то функция должна возвращать None.

Пример DataFrame:


customer_df = pd.DataFrame({
            'number': [0, 1, 2, 3, 4],
            'cust_id': [128, 1201, 9832, 4392, 7472],
            'cust_age': [13, 21, 19, 21, 60],
            'cust_sale': [0, 0, 0.2, 0.15, 0.3],
            'cust_year_birth': [2008, 2000, 2002, 2000, 1961],
            'cust_order': [1400, 14142, 900, 1240, 8430]
        })

In [28]:
import pandas as pd

def delete_columns(df, col=[]):
    # Проверяем, есть ли все указанные столбцы в DataFrame
    if not set(col).issubset(df.columns):
        return None
    # Удаляем указанные столбцы и возвращаем новый DataFrame
    return df.drop(columns=col)

# Пример DataFrame
customer_df = pd.DataFrame({
    'number': [0, 1, 2, 3, 4],
    'cust_id': [128, 1201, 9832, 4392, 7472],
    'cust_age': [13, 21, 19, 21, 60],
    'cust_sale': [0, 0, 0.2, 0.15, 0.3],
    'cust_year_birth': [2008, 2000, 2002, 2000, 1961],
    'cust_order': [1400, 14142, 900, 1240, 8430]
})

# Пример вызова функции
new_df = delete_columns(customer_df, ['number', 'cust_age'])
print(new_df)


   cust_id  cust_sale  cust_year_birth  cust_order
0      128       0.00             2008        1400
1     1201       0.00             2000       14142
2     9832       0.20             2002         900
3     4392       0.15             2000        1240
4     7472       0.30             1961        8430


Задание 2.4

Задан DataFrame countries_df, содержащий следующие столбцы: название страны, население (population) в миллионах человек и площадь страны (area) в квадратных километрах.
countries_df = pd.DataFrame({
    'country': ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан'],
    'population': [56.29, 38.05, 322.28, 146.24, 45.5, 9.5, 17.04],
    'area': [133396, 9984670, 9826630, 17125191, 603628, 207600, 2724902]
})

Для каждой страны рассчитайте плотность населения (количество человек на квадратный километр).

Затем по полученным данным рассчитайте среднее по плотностям населения в указанных странах. Ответ округлите до сотых.

Плотность населения рассчитывается как количество человек, проживающих на территории отдельной страны, делённое на площадь этой страны. Обратите внимание, что население в таблице представлено в миллионах.

In [29]:
import pandas as pd

countries_df = pd.DataFrame({
    'country': ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан'],
    'population': [56.29, 38.05, 322.28, 146.24, 45.5, 9.5, 17.04],
    'area': [133396, 9984670, 9826630, 17125191, 603628, 207600, 2724902]
})

melb_df = countries_df.copy()

# Вычисляем плотность населения (население в миллионах, поэтому * 10^6)
melb_df['density'] = (melb_df['population'] * 1e6) / melb_df['area']

# Вычисляем среднюю плотность населения и округляем до сотых
avg_density = round(melb_df['density'].mean(), 2)

print(avg_density)

84.93


Задание 3.3

Создайте в таблице melb_df признак WeekdaySale (день недели). Найдите, сколько объектов недвижимости было продано в выходные (суббота и воскресенье), результат занесите в переменную weekend_count. В качестве ответа введите результат вывода переменной weekend_count.

In [ ]:
import pandas as pd

melb_df['WeekdaySale'] = melb_df['Date'].dt.dayofweek  # 0 - понедельник, 6 - воскресенье

# Фильтруем выходные (5 - суббота, 6 - воскресенье)
weekend_count = melb_df[melb_df['WeekdaySale'].isin([5, 6])].shape[0]

# Вывод результата
print(weekend_count)

# 12822